Add a v2 provider which is active for Android M and above

timrae opened the pull request:
This PR is designed to solve two problems:

1) Improve security of the ContentProvider when the user has an Android M device or higher, as we don't need to worry about the install order dependency design flaw in the Android framework.

This is achieved by refactoring the ContentProvider code into an abstract class, and then splitting into two implementations: `CardContentProvider` (a legacy version using the previous permissions structure), and `CardContentProvider2` (a new provider that uses properly enforced static permissions, but with the same implementation as before).

On Android M and above, the legacy provider is completely disabled, and the setting in advanced preferences is removed as the Android operating system does a better job of handling access, by way of our custom permission.

2) Fix the failing unit tests

This is achieved by switching the unit tests over to `CardContentProvider2`, which makes the tests pass due to the use of static permissions.

marcardar commented:
Okay, that's a lot of new code. Do you think it's worth it?

timrae commented:
> Do you think it's worth it?

Yes, this is something we were inevitably going to have do eventually.

marcardar commented:
I would have two separate providers (i.e. no abstract superclasses). Leave the old one as the provider before all our recent changes. And then the new one.

marcardar commented:
Also, I wouldn't only use the new provider on 23+. We can also use it whenever the permission has been granted, right?

timrae commented:
> I would have two separate providers (i.e. no abstract superclass). Leave the old one as the provider before all our recent changes. And then the new one.

Nah, too much duplicate code to maintain IMO. We need to keep maintaining the legacy version for the foreseeable future as Lollipop is going to be around for a long time to come.

> I wouldn't only use the new provider on 23+. We can also use it whenever the permission has been granted, right?

Yeah, good idea. Currently it won't make any difference as the implementation is the same, but I agree it's better to use the new provider when possible! I just pushed an update with that change.

marcardar commented:
> Nah, too much duplicate code to maintain IMO. We need to keep maintaining the legacy version for the foreseeable future as Lollipop is going to be around for a long time to come.

Yeah, I agree. It's not so much "legacy" code if it's being actively developed!

Still, I'd be cautious about rushing into something like this. Have you looked into path permissions? This might be a good way of protecting notes according to their model. This is a good overview of it: https://www.owasp.org/images/c/ca/ASDC12-An_InDepth_Introduction_to_the_Android_Permissions_Modeland_How_to_Secure_MultiComponent_Applications.pdf

What we currently have, does actually work. Your PR is a good step towards a better permission system, but I can't see that it does anything that the current code doesn't already do. So, it might be worth implementing something like path permissions to make the extra complexity worthwhile.

marcardar commented:
Something I've been wondering... Please could you check if it's possible for the provider to query the calling app's Manifest for the uses-permission tag (using [PackageManager](https://developer.android.com/reference/android/content/pm/PackageInfo.html#requestedPermissions))? I'm wondering if on API <23, when the permission is not granted, maybe the provider can look for the tag (to workaround the app-install-order problem). If this works then the problem devices are on those below Kitkat.

marcardar commented:
This is also worth reading: https://code.google.com/p/android/issues/detail?id=65864

timrae commented:
I fixed two bugs, one where the provider was still using the wrong URI (causing the tests to fail), and another where the API was trying to use the new provider even for ApiHostSpecVersion 1. I also changed the legacy provider to throw SecurityException instead of IllegalStateException as this seems to be the conventional exception to use.

timrae commented:
> Have you looked into path permissions? This might be a good way of protecting notes according to their model. 

Yeah I've looked into it a little. According to the [documentation](http://developer.android.com/guide/topics/manifest/provider-element.html#gprmns):

> (Allow apps) who ordinarily would not have permission to access the content provider's data (to) be granted permission to do so, temporarily overcoming the restriction imposed by the readPermission, writePermission, and permission attributes.

I suppose it might be possible to use them in the future in addition to the changes in this PR, but they're not really applicable to the task at hand AFAIK.

> What we currently have, does actually work. Your PR is a good step towards a better permission system, but I can't see that it does anything that the current code doesn't already do.

The code in this PR locks down access to all of the provider methods (query(), insert(), update(), delete()) on Android M, completely disables the old provider, and removes the obsolete preference that re-enables it. The current code in develop only locks down update() and delete(), and the permissions are checked dynamically there, which IMO isn't as nice / obvious as the static permissions used in this new provider.

marcardar commented:
> I suppose it might be possible to use them in the future in addition to the changes in this PR, but they're > not really applicable to the task at hand AFAIK.

Yeah, I'm not sure exactly how this would work, but it's important to note you can grant persistable permissions (they persist across reboots until the app is uninstalled). I do something related to this in my app - the app asks the user to select a directory and by doing so the user is granting the app permission to write to it (to use as a backup folder). That folder can be on external storage and the app never needs to be granted WRITE_EXTERNAL_STORAGE permission.  As a side note - this is something you could look into for the AnkiDroid folder.

> The current code in develop only locks down update() and delete(), and the permissions are checked dynamically there, which IMO isn't as nice / obvious as the static permissions used in this new provider.

Adding query(), insert() checks would take less than a minute :) And this approach makes it very easy to have a strategy that says, okay, the client doesn't have the permission, but that's maybe because of the app-install-order thing so grant them read/insert rights. All in the same class.

Yeah, it's nicer to lock down the permissions in the manifest, but not 200 lines-of-code nicer.

I just wonder that with this annoying app-install-order issue, we can tackle it much better with runtime permission checks in the provider.  I've got a strong feeling the long-term solution to permission checking will involve paths/uris (Android's moving that way generally) so I think it's worth investigating that fully.

timrae commented:
> it's nicer to lock down the permissions in the manifest, but not 200 lines-of-code nicer.

ugh... I really thought this was the way to proceed but I can't really argue with that I guess. Dynamic permissions are definitely going to be simpler, and that will get rid of some slightly unpleasant code that's getting introduced into the API here. 

With dynamic permissions, what permissions related API methods would we need to give the client?

timrae commented:
I guess we need to properly document that most of our methods throw an exception if called without permission as well... Catching these will be a bit of a pain for the the clients.

marcardar commented:
I'll make some changes to my permissions PR. To confirm provider behaviour:

1. 23+: Everything protected by permission
2. <23: update/delete protected by permission, everything else unprotected

Is this right?

It might be that we don't need a canUpdate() method in the API. Documentation should be enough.

timrae commented:
> Is this right?

Yes. Also, let's throw SecurityException instead of IllegalArgumentException
